experimental/virtio/src/vsock/packet.rs

```diff
@@ -0,0 +1,299 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Virtio vsock packet implementation.
+//!
+//! A packet consists of a header and an optional payload.
+//!
+//! See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-3960006>.
+
+use alloc::vec::Vec;
+use anyhow::Context;
+use bitflags::bitflags;
+use core::mem::size_of;
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+
+/// The size of the packet header in bytes.
+const HEADER_SIZE: usize = 44;
+/// The offset to the src_cid field.
+const SRC_CID_OFFSET: usize = 0;
+/// The offset to the dst_cid field.
+const DST_CID_OFFSET: usize = 8;
+/// The offset to the src_port field.
+const SRC_PORT_OFFSET: usize = 16;
+/// The offset to the dst_port field.
+const DST_PORT_OFFSET: usize = 20;
+/// The offset to the len field.
+const LEN_OFFSET: usize = 24;
+/// The offset ot the type field.
+const TYPE_OFFSET: usize = 28;
+/// The offset ot the op field.
+const OP_OFFSET: usize = 30;
+/// The offset ot the flags field.
+const FLAGS_OFFSET: usize = 32;
+/// The offset ot the buf_alloc field.
+const BUF_ALLOC_OFFSET: usize = 36;
+/// The offset ot the fwd_cnt field.
+const FWD_CNT_OFFSET: usize = 40;
+
+pub struct Packet {
+    buffer: Vec<u8>,
+}
+
+impl Packet {
+    /// Creates a new `Packet` from an existing buffer.
+    pub fn new(buffer: Vec<u8>) -> anyhow::Result<Self> {
+        if buffer.len() < HEADER_SIZE {
+            anyhow::bail!("vsock packet too short");
+        }
+        Ok(Self { buffer })
+    }
+
+    /// Creates a new control `Packet` with only a header.
+    pub fn new_header_only() -> Self {
+        let buffer = (0..HEADER_SIZE).map(|_| 0u8).collect();
```

Comment:
Nit, but instead of creating a Range and iterating, creating a slice might be clearer: `[0u8; HEADER_SIZE].to_vec()`